src/prefect/core/flow.py

```diff
@@ -143,6 +143,12 @@ def my_task():
             the flow (e.g., presence of cycles and illegal keys) after adding the edges passed
             in the `edges` argument. Defaults to the value of `eager_edge_validation` in
             your prefect configuration file.
+        -  terminal_state_handler (Callable, optional): A state handler that mutates the final state
+            of the flow, with signature `state_handler(state: State, key_states: Set[State],
+            return_states: Dict[Task, State], terminal_states: Set[State],) -> State`.
+            `state` is the current state of the Flow. `key_states` are the states that will determine
+            the success / failure of the flow run. `return_states` contains states for Flow Tasks.
+            `terminal_states` are the states of terminal tasks for this Flow.
```

Comment:
Currently this is written in a way that exposes the internal workings of the `FlowRunner` in a way that the user probably doesn't care about (and we wouldn't want to guarantee wouldn't change). For example - `return_states` is all states that are requested to return (No states by default). `key_states` are known to users as "reference task states" (`key_states` is just an internal variable name, not a term a user would recognize). I'd argue that we want to present an API here that is easier for a user to understand without exposing internal details. We'd also want something closer to the existing api for other state handlers (for consistency).

Perhaps:

```python
def terminal_state_handler(flow: Flow, state: State, task_states: Dict[Task, State]) -> Optional[State]:
    """Determine the final state for a flow.

    Args:
        - flow (Flow): the flow itself.
        - state (State): the current flow state.
        - task_states (Dict[Task, State]): A mapping from all tasks to their
            corresponding states.
    Returns:
        - state (State, optional): Optionally return a new state if different
            than the original state. If no state is returned, the original state
            will be used.
    """
```

If a user wants to to make a decision based on reference tasks (possible) or terminal tasks (unlikely) they can query those from the `flow` object directly and then use the `task_states` mapping to map tasks to their corresponding states. But it also provides access to all included task states under a uniform and more understandable api.

Thoughts?

Reply:
I agree the simplification may be nice since it's user-facing but we'll probably want to create an example of making the decision based on reference tasks then.

Reply:
Yep this makes sense, I'll update the signature and docs accordingly

Reply:
Updated and rebased to clean up the rewrite. Let me know what you think. Is there a better example I could use for the docs?